src/sparseml/yolov8/utils/export_samples.py

```diff
@@ -0,0 +1,174 @@
+    os.makedirs(sample_out_dir_ort, exist_ok=True)
+
+    save_inputs_as_uint8 = _graph_has_uint8_inputs(onnx_path) if onnx_path else False
+
+    # Prepare model for inference
+    model = model.to(device)
+    model.eval()
+
+    # Prepare onnxruntime engine for inference
+    ort_session = onnxruntime.InferenceSession(onnx_path)
+
+    LOGGER.info(f"Exporting sample inputs to directory {sample_in_dir}")
+    LOGGER.info(f"Exporting sample torch outputs to directory {sample_out_dir_torch}")
+    LOGGER.info(
+        f"Exporting sample onnxruntime outputs to directory {sample_out_dir_ort}"
+    )
+
+    for batch in data_loader:
+        file_idx = f"{exported_samples}".zfill(4)
+        preprocessed_batch = preprocess(batch=batch, device=device)
+        image = preprocessed_batch["img"]
+
+        # Save inputs as numpy array
+        _export_inputs(image, sample_in_dir, file_idx, save_inputs_as_uint8)
+        # Save torch outputs as numpy array
+        _export_torch_outputs(image, model, sample_out_dir_torch, file_idx)
+        # Save onnxruntime outputs as numpy array
+        _export_ort_outputs(
+            image.cpu().numpy(), ort_session, sample_out_dir_ort, file_idx
+        )
+
+        exported_samples += 1
+
+        if exported_samples >= number_export_samples:
+            break
+
+    if exported_samples < number_export_samples:
+        LOGGER.info(
+            f"Could not export {number_export_samples} samples. Exhausted dataloader "
+            f"and exported {exported_samples} samples",
+            level="warning",
+        )
+
+    LOGGER.info(
+        f"Completed the export of {number_export_samples} "
+        f"input/output samples to {save_dir}"
+    )
+
+
+def _export_torch_outputs(
+    image: torch.Tensor, model: torch.nn.Module, sample_out_dir: str, file_idx: str
+):
+
+    # Run model to get torch outputs
+    model_out = model(image)
+    # preds, intermediate_outputs = model_out
+    preds, _ = model_out
+
+    # Move to cpu for exporting
+    preds = preds.detach().to("cpu")
+
+    sample_output_filename = os.path.join(sample_out_dir, f"inp-{file_idx}.npz")
+    numpy.savez(sample_output_filename, preds)
+
+
+def _export_ort_outputs(
+    image: numpy.ndarray,
+    session: onnxruntime.InferenceSession,
+    sample_out_dir: str,
+    file_idx: str,
+):
+
+    # Run model to get onnxruntime outputs
+    ort_inputs = {session.get_inputs()[0].name: image}
+    ort_outs = session.run(None, ort_inputs)
+    # preds, iterm_out1, interm_out2, interm_out3 = model_out
+    preds, *_ = ort_outs
+
+    sample_output_filename = os.path.join(sample_out_dir, f"inp-{file_idx}.npz")
+    numpy.savez(sample_output_filename, preds)
+
+
+def _export_inputs(
+    image: torch.Tensor, sample_in_dir: str, file_idx: str, save_inputs_as_uint8: bool
+):
+
+    sample_in = image.detach().to("cpu")
+    if save_inputs_as_uint8:
+        sample_in = (255 * sample_in).to(dtype=torch.uint8)
```

Comment:
What if inputs are already quantized? Could that condition occur?

Reply:
Inputs come from the yolov8 data loader. You could in theory pass the argument so that their represent is reduced to `half` (FP16). I could account for that, but maybe as a separate feature.